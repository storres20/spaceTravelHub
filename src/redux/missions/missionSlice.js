/* eslint-disable camelcase */
/* eslint-disable max-len */
// missionSlice.js
import { createAsyncThunk } from '@reduxjs/toolkit';

// Actions... types
const GET_MISSIONS = 'spaceTravelHub/missions/GET_MISSIONS';
const JOIN_MISSIONS = 'spaceTravelHub/missions/JOIN_MISSIONS';

// Reducer
export default function reducer(state = [], action) {
  switch (action.type) {
    // do reducer stuff
    // GET missions from the API
    case `${GET_MISSIONS}/fulfilled`:
      return action.payload;
    // Implement mission joining
    case JOIN_MISSIONS: {
      const newState = state.map((mission) => {
        if (mission.mission_id !== action.payload) {
          return mission;
        } if (mission.reserved) {
          return { ...mission, reserved: false };
        }
        return { ...mission, reserved: true };
      });
      return newState; }
    default: return state;
  }
}

// Action Creators
export const getMissions = createAsyncThunk(GET_MISSIONS, async () => {
  const getMissionsUrl = 'https://api.spacexdata.com/v3/missions';
  const response = await fetch(getMissionsUrl,
    {
      method: 'GET',
      headers: {
        'content-Type': 'application/json',
      },
    });
  const result = await response.json();

  /* Keep some keys from Array of Objects and remove the rest of them */
  /* 'mission_id', 'mission_name', 'description'; */

  const redux = result.map(({ mission_id, mission_name, description }) => ({ mission_id, mission_name, description }));

  return redux;
}); /* getMissions - createAsyncThunk - API */

export function joinMissions(obj) {
  return { type: JOIN_MISSIONS, payload: obj };
} /* joinMissions */
