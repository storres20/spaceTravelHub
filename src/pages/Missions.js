/* eslint-disable camelcase */
import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import Table from 'react-bootstrap/Table'; /* React Bootstrap */
import Button from 'react-bootstrap/Button'; /* React Bootstrap */
import Badge from 'react-bootstrap/Badge'; /* React Bootstrap */
import { getMissions } from '../redux/missions/missionSlice';

function Missions() {
  /* Setup Redux dispatch */
  const dispatch = useDispatch();

  /* At the beginning missions => [], after dispatch missions => [{},{}...{}] */
  const missions = useSelector((state) => state.missions);

  /* For dispatching only once */
  useEffect(() => {
    dispatch(getMissions());
  }, [dispatch]);

  return (
    <Table striped bordered hover className="mt-3">
      <thead>
        <tr>
          <th>Mission</th>
          <th>Description</th>
          <th>Status</th>
          <th> </th>
        </tr>
      </thead>
      <tbody>
        {missions.map((item) => (
          <tr key={item.mission_id}>
            <td><b>{item.mission_name}</b></td>
            <td>{item.description}</td>
            <td className="align-middle text-center">
              <Badge bg={item.reserved ? 'info' : 'secondary'}>
                {item.reserved ? 'Active Member' : 'NOT A MEMBER'}
              </Badge>
            </td>
            <td className="align-middle text-center" style={{ width: '12%' }}>
              <Button variant={item.reserved ? 'outline-danger' : 'outline-secondary'}>
                {item.reserved ? 'Leave Mission' : 'Join Mission'}
              </Button>
            </td>
          </tr>
        ))}

      </tbody>
    </Table>
  );
}

export default Missions;
