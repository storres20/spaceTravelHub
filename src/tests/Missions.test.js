import React from 'react';
import renderer from 'react-test-renderer';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';
import Missions from '../pages/Missions';
// import moviesReducer from '../redux/movies/movies';
import missions from '../redux/missions/missionSlice';

// configureStore
const createTestStore = () => {
  const store = configureStore({
    reducer: {
      missions,
    },
  });

  return store;
};
export default createTestStore;

// Test
describe('My Connected React-Redux Component', () => {
  let store;
  let component;

  beforeEach(() => {
    store = createTestStore();

    // Create a redux store
    component = renderer.create(
      <Provider store={store}>
        <Missions />
      </Provider>,
    ).toJSON();
  });

  test('renders correctly', () => {
    expect(component).toMatchSnapshot();
  });
});
