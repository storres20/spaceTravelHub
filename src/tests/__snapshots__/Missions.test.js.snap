// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`My Connected React-Redux Component renders correctly 1`] = `
<table
  className="mt-3 table table-striped table-bordered table-hover"
>
  <thead>
    <tr>
      <th>
        Mission
      </th>
      <th>
        Description
      </th>
      <th>
        Status
      </th>
      <th>
         
      </th>
    </tr>
  </thead>
  <tbody />
</table>
`;
